plugins/kosync.koplugin/main.lua

```diff
@@ -24,27 +25,86 @@ local KOSync = WidgetContainer:extend{
     is_doc_only = true,
     title = _("Register/login to KOReader server"),
 
-    page_update_times = 0,
-    last_page = -1,
-    last_page_turn_ticks = 0,
+    push_timestamp = nil,
+    pull_timestamp = nil,
+    page_update_counter = nil,
+    last_page = nil,
+    last_page_turn_timestamp = nil,
+    periodic_push_task = nil,
+    periodic_push_scheduled = nil,
+
+    settings = nil,
 }
 
 local SYNC_STRATEGY = {
-    -- Forward and backward whisper sync settings are using different
-    -- default value, so none of following opinions should be zero.
-    PROMPT = 1,
-    WHISPER = 2,
+    PROMPT  = 1,
+    SILENT  = 2,
     DISABLE = 3,
-
-    DEFAULT_FORWARD = 1,
-    DEFAULT_BACKWARD = 3,
 }
 
 local CHECKSUM_METHOD = {
     BINARY = 0,
     FILENAME = 1
 }
 
+-- Debounce push/pull attempts
+local API_CALL_DEBOUNCE_DELAY = time.s(25)
+
+-- NOTE: This is used in a migration script by ui/data/onetime_migration,
+--       which is why it's public.
+KOSync.default_settings = {
+    custom_server = nil,
+    username = nil,
+    userkey = nil,
+    -- Do *not* default to auto-sync on devices w/ NetworkManager support, as wifi is unlikely to be on at all times there, and the nagging enabling this may cause requires careful consideration.
+    auto_sync = not Device:hasWifiManager(),
+    pages_before_update = nil,
+    sync_forward = SYNC_STRATEGY.PROMPT,
+    sync_backward = SYNC_STRATEGY.DISABLE,
+    checksum_method = CHECKSUM_METHOD.BINARY,
+}
+
+function KOSync:init()
+    self.push_timestamp = 0
+    self.pull_timestamp = 0
+    self.page_update_counter = 0
+    self.last_page = -1
+    self.last_page_turn_timestamp = 0
+    self.periodic_push_scheduled = false
+
+    -- Like AutoSuspend, we need an instance-specific task for scheduling/resource management reasons.
+    self.periodic_push_task = function()
+        self.periodic_push_scheduled = false
+        self.page_update_counter = 0
+        -- We do *NOT* want to make sure networking is up here, as the nagging would be extremely annoying; we're leaving that to the network activity check...
+        self:updateProgress(false, false)
+    end
+
+    self.settings = G_reader_settings:readSetting("kosync", self.default_settings)
+    self.device_id = G_reader_settings:readSetting("device_id")
+
+    -- Disable auto-sync if beforeWifiAction was reset to "prompt" behind our back...
+    if self.settings.auto_sync and Device:hasWifiManager() and G_reader_settings:readSetting("wifi_enable_action") ~= "turn_on" then
+        self.settings.auto_sync = false
+        logger.warn("KOSync: Automatic sync has been disabled because wifi_enable_action is *not* turn_on")
+    end
+
+    self.ui.menu:registerToMainMenu(self)
+end
+
+function KOSync:getSyncPeriod()
+    if not self.settings.auto_sync then
+        return _("Unavailable")
```

Comment:
Probably better to use "Not available" (which is already translated), unless there's a specific reason to use unavailable.

Reply:
Not particularly, no ;).